libpromises/dbm_lmdb.c

```diff
@@ -490,6 +533,220 @@ static void HandleLMDBCorruption(MDB_env *env, const char *msg)
+        close(fd_tstamp);
+        free(db_lock_file);
+        free(tstamp_file);
+
+        exit(EC_CORRUPTION_REPAIR_FAILED);
+    }
+    FileLock db_lock = { .fd = fd_db_lock };
+
+    int ret;
+    bool handle_rotation = true;
+
+    /* Make sure we are not holding the DB's lock (potentially needed by some
+     * other process for the repair or rotation) to avoid deadlocks. */
+    Log(LOG_LEVEL_DEBUG, "Releasing lock on the '%s' DB", lmdb_file);
+    ExclusiveFileUnlock(&db_lock, false); /* close=false */
+
+    ret = SharedFileLock(&tstamp_lock, true);
+    if (ret == 0)
+    {
+        if (RotatedAfterOpen(lmdb_file, fd_tstamp))
+        {
+            /* The corruption has already been handled. This process should just
+             * die because we have no way to return to the point where it would
+             * just open the new (repaired or rotated) LMDB file. */
+            handle_rotation = false;
+        }
+        SharedFileUnlock(&tstamp_lock, false);
+    }
+    else
+    {
+        /* should never happen (we tried to wait), but if it does, just log an
+         * error and keep going */
+        Log(LOG_LEVEL_ERR,
+            "Failed to get shared lock for the rotation timestamp of the '%s' DB",
+            lmdb_file);
+    }
+
+    if (!handle_rotation)
+    {
+        /* Just clean after ourselves and terminate the process. */
+        ThreadUnlock(cft_db_corruption_lock);
+        close(fd_db_lock);
+        close(fd_tstamp);
+        free(db_lock_file);
+        free(tstamp_file);
+
+        exit(EC_CORRUPTION_REPAIRED);
+    }
+
+    /* HERE is a window for some other process to do the rotation between when we
+     * checked the timestamp using the shared lock above and the attempt to get
+     * the exclusive lock right below. However, this is detected by checking the
+     * contents of the timestamp file again below, while holding the EXCLUSIVE
+     * lock. */
+
+    ret = ExclusiveFileLock(&tstamp_lock, true);
+    if (ret != 0)
+    {
+        /* should never happen (we tried to wait), but if it does, just
+         * terminate because doing the rotation without the lock could be
+         * disasterous */
+        Log(LOG_LEVEL_ERR,
+            "Failed to get shared lock for the rotation timestamp of the '%s' DB",
+            lmdb_file);
+
+        ThreadUnlock(cft_db_corruption_lock);
+        close(fd_db_lock);
+        close(fd_tstamp);
+        free(db_lock_file);
+        free(tstamp_file);
+
+        exit(EC_CORRUPTION_REPAIR_FAILED);
+    }
+
+    /* Cleared to resolve the corruption. */
+
+    /* 1. Acquire the lock for the DB to prevent more processes trying to use
+     *    it while it is corrupted (wait till the lock is available). */
+    while (ExclusiveFileLock(&db_lock, false) == -1)
+    {
+        /* busy wait to do the logging */
+        Log(LOG_LEVEL_INFO, "Waiting for the lock on the '%s' DB",
+            lmdb_file);
+        sleep(1);
+    }
```

Comment:
Forever?

Reply:
Yup. If this gives up, the next instance of the same program will again end up here anyway. I guess I could make the log level increment over time to make this easier to detect, but that's as much as we can do, AFAICT.